src/tdsops.f90

```diff
@@ -0,0 +1,700 @@
+         self%dist_sa(n-3) = 2._dp/11._dp
+         self%dist_sc(n-3) = 2._dp/11._dp
+         self%coeffs_b(1, :) = self%coeffs_b(2, :)
+      end select
+
+      call self%preprocess(dist_b)
+
+   end subroutine deriv_2nd
+
+   subroutine interpl_mid(self, variation, from_to, bc_top, bc_bot, paire)
+      implicit none
+
+      class(tdsops_t), intent(inout) :: self
+      character(*), intent(in) :: variation, from_to
+      character(*), optional, intent(in) :: bc_top, bc_bot
+      logical, optional, intent(in) :: paire
+
+      real(dp), allocatable :: dist_b(:)
+      real(dp) :: aici, bici, cici, dici
+      integer :: i, n, n_halo
+
+      ! alpha is ailcai
+
+      select case(variation)
+      case('classic')
+         self%alpha = 0.3_dp
+         aici = 0.75_dp
+         bici = 0.05_dp
+         cici = 0._dp
+         dici = 0._dp
+      case('optimised')
+         self%alpha = 0.461658_dp
+         dici = 0.00146508_dp
+         aici = (75._dp + 70._dp*self%alpha - 640._dp*dici)/128._dp
+         bici = (-25._dp + 126._dp*self%alpha + 2304._dp*dici)/256._dp
+         cici = (3._dp - 10._dp*self%alpha - 1280._dp*dici)/256._dp
+      case('aggressive')
+         self%alpha = 0.49_dp
+         aici = (75._dp + 70._dp*self%alpha)/128._dp
+         bici = (-25._dp + 126._dp*self%alpha)/256._dp
+         cici = (3._dp - 10._dp*self%alpha)/256._dp
+         dici = 0._dp
+      case default
+         error stop 'variation is not defined'
+      end select
+
+      select case(from_to)
+      case('v2p')
+         self%coeffs(:) = [0._dp, dici, cici, bici, &
+                           aici, &
+                           aici, bici, cici, dici]
+      case('p2v')
+         self%coeffs(:) = [dici, cici, bici, aici, &
+                           aici, &
+                           bici, cici, dici, 0._dp]
+      end select
+
+      do i = 1, self%n_halo
+         self%coeffs_t(i,:) = self%coeffs(:)
+         self%coeffs_b(i,:) = self%coeffs(:)
+      end do
+
+      self%dist_sa(:) = self%alpha; self%dist_sc(:) = self%alpha
+
+      n = self%n
+      n_halo = self%n_halo
+
+      allocate(dist_b(n))
+      dist_b(:) = 1._dp
+
+      if (bc_top /= 'periodic' .or. bc_top /= 'null') then
+         self%dist_sa(1) = 0._dp
+
+         select case(from_to)
+         case('v2p')
+            ! npaire is always 1
+            dist_b(1) = 1._dp + self%alpha
+            self%coeffs_t(1, :) = [0._dp, 0._dp, 0._dp, 0._dp, &
+                                   aici, &
+                                   aici + bici, bici + cici, cici + dici, dici]
+            self%coeffs_t(2, :) = [0._dp, 0._dp, 0._dp, bici, &
+                                   aici + cici, &
+                                   aici + dici, bici, cici, dici]
+            self%coeffs_t(3, :) = [0._dp, 0._dp, cici, bici + dici, &
+                                   aici, &
+                                   aici, bici, cici, dici]
+         case('p2v')
+            ! npaire is always 1
```

Comment:
I am not sure about npaire=1 if the user wants to interpolate the pressure on the velocity grid instead of computing the pressure gradient.